bundle/config/mutator/run_as.go

```diff
@@ -23,12 +27,120 @@ func (m *setRunAs) Name() string {
+var allowListForRunAsOther = []string{"jobs", "models", "registered_models", "experiments"}
+
+// Resources that do not allow setting a run_as identity to a different user but
+// have plans to add platform side support for `run_as` for these resources at
+// some point in the future. For example, pipelines, model serving endpoints or lakeview dashboards.
+//
+// We expect the allow list and the deny list to form mutually exclusive and exhaustive
+// sets of all resource types that are supported by DABs.
+var denyListForRunAsOther = []string{"pipelines", "model_serving_endpoints"}
+
+type errorUnsupportedResourceTypeForRunAs struct {
+	resourceType  string
+	resourceValue dyn.Value
+	currentUser   string
+	runAsUser     string
+}
+
+// TODO(6 March 2024): This error message is big. We should split this once
+// diag.Diagnostics is ready.
+// TODO(6 March 2024): Link the docs page describing run_as semantics in the error below
+// once the page is ready.
+func (e errorUnsupportedResourceTypeForRunAs) Error() string {
+	return fmt.Sprintf("%s are not supported when the current deployment user is different from the bundle's run_as identity. Please deploy as the run_as identity. List of supported resources: [%s]. Location of the unsupported resource: %s. Current identity: %s. Run as identity: %s", e.resourceType, strings.Join(allowListForRunAsOther, ", "), e.resourceValue.Location(), e.currentUser, e.runAsUser)
+}
+
+func getRunAsIdentity(runAs dyn.Value) (string, error) {
+	// Get service principal name and user name from run_as section
+	runAsSp, err := dyn.Get(runAs, "service_principal_name")
+	if err != nil && !dyn.IsNoSuchKeyError(err) {
+		return "", err
+	}
+	runAsUser, err := dyn.Get(runAs, "user_name")
+	if err != nil && !dyn.IsNoSuchKeyError(err) {
+		return "", err
+	}
+
+	sp, spIsDefined := runAsSp.AsString()
+	user, userIsDefined := runAsUser.AsString()
+
+	switch {
+	case spIsDefined && userIsDefined:
+		return "", fmt.Errorf("run_as section must specify exactly one identity. A service_principal_name %q is specified at %s. A user_name %q is defined at %s", sp, runAsSp.Location(), user, runAsUser.Location())
+	case spIsDefined:
+		return sp, nil
+	case userIsDefined:
+		return user, nil
+	default:
+		return "", nil
+	}
+}
+
 func (m *setRunAs) Apply(_ context.Context, b *bundle.Bundle) error {
+	// Return early if run_as is not defined in the bundle
 	runAs := b.Config.RunAs
 	if runAs == nil {
 		return nil
 	}
 
+	currentUser := b.Config.Workspace.CurrentUser.UserName
+
+	// Assert the run_as configuration is valid in the context of the bundle
+	err := b.Config.Mutate(func(v dyn.Value) (dyn.Value, error) {
+		// Get run_as from the bundle
+		runAs, err := dyn.Get(v, "run_as")
+
+		// If run_as is not defined in the bundle, return early
+		if dyn.IsNoSuchKeyError(err) {
+			return v, nil
+		}
+		if err != nil {
+			return dyn.InvalidValue, err
+		}
+
+		runAsIdentity, err := getRunAsIdentity(runAs)
```

Comment:
Can't you just read from `b.Config.Bundle.RunAs` directly to simplify things? Would it be an anti-pattern or something? cc @pietern

Reply:
The tradeoffs I see here are
pros:
1. The code is simpler and easier to read
2. The mutator consistently uses dynamic typing for configuration processing.

cons:
1. Location information is missing in the error messages.
2. Harder to reason about. Safety guarentees from strongly typed objects is missing and thus testing and code needs to me more through / verbose.

One reason I went with the approach I did is as an exercise for using the dynamic configuration more in our mutators as it'll become a bigger part of DABs. Specially, since all the `dyn` APIs are new and not used as getter / setter APIs in our mutators. 

I think it's worth iterating on this again and moving more functionality down the stack. For example on the next iteration we could add methods like `Filter` or `Match` or `GetOneOf(...)` to dyn and simplify the use of `dyn` here. I'd be interested to take another pass at it and keep using `dyn` but make the mutators more clean by moving the complexity down the stack. I'm also interested in what @pietern has to say here.

Reply:
I don't see reading from the typed structs directly as an anti-pattern. It will often be more straightforward and easy to read compared to going the `dyn` route. I'd say anywhere where you need access to the location information for error messages, where you need to deal with zero-values, variables, or have something like dynamic traversal of the configuration tree (where you'd have to jump through hoops to make it work with types), use `dyn`. For everything else, we can continue to use the existing typed approach.

Reply:
Also agree on moving functionality into `dyn` where it makes sense. It must be generically applicable though. If it is only applicable in a bundle context we can use helpers in the bundle package to share functionality.

Reply:
Moved to using typed configuration as much as possible. That indeed seems to have made the code easier to reason about for devs with fresh eyes. Moving forward defaulting to the typed configuration seems like the better option.

Added function to get location from the typed config.